python/kserve/kserve/model_server.py

```diff
@@ -264,86 +253,77 @@ def __init__(
+            self.http_port,
             # By setting log_config to None we tell Uvicorn not to configure logging as it is already
             # configured by kserve.
             log_config=None,
-            access_log_format=self.access_log_format,
-            workers=self.workers,
+            access_log_format=access_log_format,
+            workers=workers,
         )
-        await self._rest_server.run()
-
-    def start(self, models: List[BaseKServeModel]) -> None:
-        """Start the model server with a set of registered models.
+        self._grpc_server = None
+        if self.enable_grpc:
+            self._grpc_server = GRPCServer(
+                grpc_port,
+                self.dataplane,
+                self.model_repository_extension,
+                kwargs=vars(args),
+            )
 
-        Args:
-            models: a list of models to register to the model server.
-        """
-        if isinstance(models, list):
-            at_least_one_model_ready = False
-            for model in models:
-                if isinstance(model, BaseKServeModel):
-                    if model.ready:
-                        at_least_one_model_ready = True
-                        self.register_model(model)
-                        # pass whether to log request latency into the model
-                        model.enable_latency_logging = self.enable_latency_logging
-                    model.start()
-                else:
-                    raise RuntimeError("Model type should be 'BaseKServeModel'")
-            if not at_least_one_model_ready and models:
-                raise NoModelReady(models)
+    def setup_event_loop(self):
+        loop = asyncio.get_event_loop()
+        if self._custom_exception_handler is None:
+            loop.set_exception_handler(self.default_exception_handler)
         else:
-            raise RuntimeError("Unknown model collection type")
+            loop.set_exception_handler(self._custom_exception_handler)
 
         if self.max_asyncio_workers is None:
             # formula as suggest in https://bugs.python.org/issue35279
             self.max_asyncio_workers = min(32, utils.cpu_count() + 4)
         logger.info(f"Setting max asyncio worker threads as {self.max_asyncio_workers}")
-        asyncio.get_event_loop().set_default_executor(
+        loop.set_default_executor(
             concurrent.futures.ThreadPoolExecutor(max_workers=self.max_asyncio_workers)
         )
 
-        async def servers_task():
-            servers = [self._serve_rest()]
-            if self.enable_grpc:
-                servers.append(self._grpc_server.start(self.max_threads))
-            await asyncio.gather(*servers)
+    def register_signal_handler(self):
+        if sys.platform == "win32":
+            sig_list = [signal.SIGINT, signal.SIGTERM, signal.SIGBREAK]
+        else:
+            sig_list = [signal.SIGINT, signal.SIGTERM, signal.SIGQUIT]
+
+        for sig in sig_list:
+            signal.signal(
```

Comment:
https://github.com/encode/uvicorn/pull/1600#issuecomment-1319933056. Using `loop.add_signal_handler` leads to the handler being called twice as uvicorn re raises the signal.